File: Lista01/PY/ex11.py
from abc import ABC, abstractmethod

# Definindo a classe abstrata Funcionario
class Funcionario(ABC):
    
    @abstractmethod
    def calcular_salario(self) -> float:
        """Método abstrato para calcular o salário"""
        pass

# Implementando a classe FuncionarioHorista
class FuncionarioHorista(Funcionario):
    def __init__(self, horas_trabalhadas: float, valor_hora: float):
        self.horas_trabalhadas = horas_trabalhadas
        self.valor_hora = valor_hora
    
    def calcular_salario(self) -> float:
        """Calcula o salário com base nas horas trabalhadas e no valor da hora"""
        return self.horas_trabalhadas * self.valor_hora

# Implementando a classe FuncionarioAssalariado
class FuncionarioAssalariado(Funcionario):
    def __init__(self, salario_mensal: float):
        self.salario_mensal = salario_mensal
    
    def calcular_salario(self) -> float:
        """Retorna o salário mensal fixo"""
        return self.salario_mensal

# Exemplo de uso
def main():
    # Criando um funcionário horista
    funcionario_horista = FuncionarioHorista(horas_trabalhadas=160, valor_hora=20)
    print(f"Salário do funcionário horista: R$ {funcionario_horista.calcular_salario():.2f}")
    
    # Criando um funcionário assalariado
    funcionario_assalariado = FuncionarioAssalariado(salario_mensal=5000)
    print(f"Salário do funcionário assalariado: R$ {funcionario_assalariado.calcular_salario():.2f}")

if __name__ == "__main__":
    main()

File: Lista01/GO/ex11.go
package main

import "fmt"

// Definindo a interface Funcionario (equivalente a uma classe abstrata)
type Funcionario interface {
	CalcularSalario() float64
}

// Definindo a struct FuncionarioHorista
type FuncionarioHorista struct {
	nome          string
	horasTrabalhadas int
	valorPorHora   float64
}

// Implementando o método CalcularSalario para FuncionarioHorista
func (f FuncionarioHorista) CalcularSalario() float64 {
	return float64(f.horasTrabalhadas) * f.valorPorHora
}

// Definindo a struct FuncionarioAssalariado
type FuncionarioAssalariado struct {
	nome     string
	salarioFixo float64
}

// Implementando o método CalcularSalario para FuncionarioAssalariado
func (f FuncionarioAssalariado) CalcularSalario() float64 {
	return f.salarioFixo
}

// Função para exibir salários
func ExibirSalario(f Funcionario) {
	fmt.Printf("Salário: R$ %.2f\n", f.CalcularSalario())
}

func main() {
	// Criando um FuncionarioHorista
	horista := FuncionarioHorista{
		nome:            "Carlos",
		horasTrabalhadas: 160,
		valorPorHora:    50.0,
	}

	// Criando um FuncionarioAssalariado
	assalariado := FuncionarioAssalariado{
		nome:        "Ana",
		salarioFixo: 7000.0,
	}

	// Exibindo os salários
	fmt.Printf("Funcionario: %s\n", horista.nome)
	ExibirSalario(horista)

	fmt.Printf("Funcionario: %s\n", assalariado.nome)
	ExibirSalario(assalariado)
}
